cli/azd/internal/telemetry/telemetry_test.go

```diff
@@ -0,0 +1,78 @@
+package telemetry
+
+import (
+	"context"
+	"sync"
+	"testing"
+
+	"github.com/azure/azure-dev/cli/azd/internal"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestGetTelemetrySystem(t *testing.T) {
+	tests := []struct {
+		name               string
+		azdVersion         string
+		expectNil          bool
+		instrumentationKey string
+	}{
+		{"DevVersion", "0.0.0-dev.0 (commit 0000000000000000000000000000000000000000)", false, devInstrumentationKey},
+		{"ProdVersion", "1.0.0 (commit 13ec2b11aa755b11640fa16b8664cb8741d5d300)", true, prodInstrumentationKey},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			orig := internal.Version
+			defer func() { internal.Version = orig }()
+			internal.Version = tt.azdVersion
+
+			ts := GetTelemetrySystem()
+
+			if tt.expectNil {
+				assert.Nil(t, ts)
+			} else {
+				require.NotNil(t, ts)
+				assert.Equal(t, tt.instrumentationKey, ts.instrumentationKey)
+				assert.NotNil(t, ts.GetTelemetryQueue())
+				assert.NotNil(t, ts.NewUploader(true))
+
+				err := ts.Shutdown(context.Background())
+				assert.NoError(t, err)
+			}
+
+			once = sync.Once{}
```

Comment:
Took me a bit to realize that this is resetting some package global state so that when we call GetTelemetrySystem for the next test it will be able to re initialize the singleton.

Is there a reason we go out of our way to have GetTelemetrySystem return a singleton value?  I'm guessing it's because the initialization is expensive, but I'm having a hard time understanding how this will get used.  For example, would this end up as something we'd end up putting in to the context.Context we flow around?

Reply:
#483 has the end-to-end changes in draft form.

The only two usages are:
1. In `main`, where we want to initialize as soon as possible for all systems (cmd pre-parsing, cmd post-parsing or otherwise) to be able to access telemetry if needed. We also want to shutdown as late as possible to ensure all telemetry is collected.
2. In the background upload process. This is done for the command to be able to get a properly constructed storage queue that points to the telemetry directory.

We would not expect any app code to need access to the singleton, so at least the number of object references is small. The app code uses the `otel` library `GetTracer()` that returns a `NoOptTracerProvider` by default if we did not initialize `GetSystem()` which is only done in `main`.

Overall, we can move away from the package global state / hard singleton state if we can construct the object in `main` and be able to provide well-constructed objects for usage. This is somewhat possible with Wallace's change, and if we continue improving on IoC/DI. We can defer the work a little until then since the tech debt here is minimal (2 object references).